File: 3955-valid-binary-strings-with-cost-limit/3955-valid-binary-strings-with-cost-limit.py
class Solution:
    def generateValidStrings(self, n: int, k: int) -> list[str]:
        ans = []
        
        def dfs(i, cost, current_str):
            if cost > k:
                return
            if i == n:
                ans.append(current_str)
                return
                
            dfs(i + 1, cost, current_str + "0")
            
            if not current_str or current_str[-1] == '0':
                dfs(i + 1, cost + i, current_str + "1")
                
        dfs(0, 0, "")
        return ans

File: 3955-valid-binary-strings-with-cost-limit/3955-valid-binary-strings-with-cost-limit.cpp
class Solution {
public:
    vector<string> generateValidStrings(int n, int k) {
        vector<string> ans;
        string cur;

        function<void(int, int, bool)> dfs = [&](int i, int cost, bool prevOne) {
            if (i == n) {
                ans.push_back(cur);
                return;
            }

    
            cur.push_back('0');
            dfs(i + 1, cost, false);
            cur.pop_back();


            if (!prevOne && cost + i <= k) {
                cur.push_back('1');
                dfs(i + 1, cost + i, true);
                cur.pop_back();
            }
        };

        dfs(0, 0, false);
        return ans;
    }
};